File: CK3toEU4/Source/CK3World/CoatsOfArms/Emblem.cpp
#include "Emblem.h"
#include "Log.h"
#include "ParserHelpers.h"

CK3::Emblem::Emblem(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void CK3::Emblem::registerKeys()
{
	registerKeyword("texture", [this](const std::string& unused, std::istream& theStream) {
		texture = commonItems::singleString(theStream).getString();
	});
	registerKeyword("color1", [this](const std::string& unused, std::istream& theStream) {
		try
		{
			color1 = laFabricaDeColor.getColor(theStream);
		}
		catch (std::exception& e)
		{
			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
			color1 = commonItems::Color(std::array<int, 3>{0, 0, 0});
		}
	});
	registerKeyword("color2", [this](const std::string& unused, std::istream& theStream) {
		try
		{
			color2 = laFabricaDeColor.getColor(theStream);
		}
		catch (std::exception& e)
		{
			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
			color2 = commonItems::Color(std::array<int, 3>{0, 0, 0});
		}
	});
	registerKeyword("color3", [this](const std::string& unused, std::istream& theStream) {
		try
		{
			color3 = laFabricaDeColor.getColor(theStream);
		}
		catch (std::exception& e)
		{
			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
			color3 = commonItems::Color(std::array<int, 3>{0, 0, 0});
		}
	});
	registerKeyword("mask", [this](const std::string& unused, std::istream& theStream) {
		mask = commonItems::intList(theStream).getInts();
	});
	registerKeyword("instance", [this](const std::string& unused, std::istream& theStream) {
		auto instance = EmblemInstance(theStream);
		if (instance.getPosition().empty())
			instance.defaultPosition();
		instances.emplace_back(instance);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: CK3ToEU4/Source/CK3/CoatsOfArms/Emblem.cs
using System;
using System.Collections.Generic;
using commonItems;
using commonItems.Colors;

namespace CK3ToEU4.CK3.CoatsOfArms;

public class Emblem
{
    public Emblem(BufferedReader reader, ColorFactory colorFactory)
    {
        var parser = new Parser();
        RegisterKeys(parser, colorFactory);
        parser.ParseStream(reader);
    }

    private void RegisterKeys(Parser parser, ColorFactory colorFactory)
    {
        parser.RegisterKeyword("texture", reader => {
            Texture = reader.GetString();
        });
        parser.RegisterKeyword("color1", reader => {
            try
            {
                Color1 = colorFactory.GetColor(reader);
            }
            catch (Exception e)
            {
                Logger.Warn(e.Message + " - sidestepping with black.");
                Color1 = new Color(0, 0, 0);
            }
        });
        parser.RegisterKeyword("color2", reader => {
            try
            {
                Color2 = colorFactory.GetColor(reader);
            }
            catch (Exception e)
            {
                Logger.Warn(e.Message + " - sidestepping with black.");
                Color2 = new Color(0, 0, 0);
            }
        });
        parser.RegisterKeyword("color3", reader => {
            try
            {
                Color3 = colorFactory.GetColor(reader);
            }
            catch (Exception e)
            {
                Logger.Warn(e.Message + " - sidestepping with black.");
                Color3 = new Color(0, 0, 0);
            }
        });
        parser.RegisterKeyword("mask", reader =>
        {
            mask = reader.GetInts();
        });
        parser.RegisterKeyword("instance", reader => {
            var instance = new EmblemInstance(reader);
            if (instance.Position.Count == 0)
                instance.DefaultPosition();
            instances.Add(instance);
        });
       parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
    }
    
    public string? Texture { get; private set; }
    public Color? Color1 { get; private set; }
    public Color? Color2 { get; private set; }
    public Color? Color3 { get; private set; }
    public IReadOnlyList<int> Mask => mask;
    private List<int> mask = [];
    public IReadOnlyList<EmblemInstance> Instances => instances;
    private readonly List<EmblemInstance> instances = [];
}